File: sdk/python/promptsentinel.py
"""PromptSentinel Python SDK — stdlib only, no extra deps."""
from __future__ import annotations

import json
import urllib.error
import urllib.request
from typing import Any, Optional


class PromptSentinelError(Exception):
    def __init__(self, status: int, message: str) -> None:
        super().__init__(f"HTTP {status}: {message}")
        self.status = status


class PromptSentinelClient:
    """Minimal client for the PromptSentinel API.

    Parameters
    ----------
    base_url:
        Base URL of your PromptSentinel instance, e.g. ``"http://localhost:8000"``.
    api_key:
        Optional API key sent as ``X-API-Key``.  Required for authenticated
        endpoints (guard scan, usage, etc.).
    """

    def __init__(self, base_url: str, api_key: Optional[str] = None) -> None:
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _headers(self) -> dict[str, str]:
        h = {"Content-Type": "application/json", "Accept": "application/json"}
        if self.api_key:
            h["X-API-Key"] = self.api_key
        return h

    def _request(self, method: str, path: str, body: Any = None) -> Any:
        url = self.base_url + path
        data = json.dumps(body).encode() if body is not None else None
        req = urllib.request.Request(url, data=data, headers=self._headers(), method=method)
        try:
            with urllib.request.urlopen(req) as resp:
                return json.loads(resp.read())
        except urllib.error.HTTPError as exc:
            try:
                detail = json.loads(exc.read()).get("detail", exc.reason)
            except Exception:
                detail = exc.reason
            raise PromptSentinelError(exc.code, detail) from exc

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def guard_scan(
        self,
        input: str,
        output: Optional[str] = None,
        context: Optional[str] = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        """Run a guard scan and return the full response dict.

        Extra keyword arguments are forwarded to the request body (e.g.
        ``policy={"block_pii": False}``).
        """
        payload: dict[str, Any] = {"input": input}
        if output is not None:
            payload["output"] = output
        if context is not None:
            payload["context"] = context
        payload.update(kwargs)
        return self._request("POST", "/guard/scan", payload)

    def get_usage_summary(self) -> dict[str, Any]:
        """Return the current billing-period usage summary."""
        return self._request("GET", "/usage/summary")

    def get_trust_score(self, days: int = 30) -> dict[str, Any]:
        """Return the organisation trust score / maturity index.

        Parameters
        ----------
        days:
            Look-back window in days (default 30).
        """
        return self._request("GET", f"/analytics/trust-score?days={days}")

File: sdk/js/promptsentinel.js
/**
 * PromptSentinel JS SDK — zero dependencies, works in Node ≥18 and browsers.
 */

export class PromptSentinelError extends Error {
  /** @param {number} status @param {string} message */
  constructor(status, message) {
    super(`HTTP ${status}: ${message}`);
    this.name = "PromptSentinelError";
    this.status = status;
  }
}

export class PromptSentinelClient {
  /**
   * @param {string} baseUrl  Base URL, e.g. "http://localhost:8000"
   * @param {string} [apiKey] X-API-Key for authenticated endpoints
   */
  constructor(baseUrl, apiKey) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey ?? null;
  }

  // ── internal ────────────────────────────────────────────────────────────

  _headers() {
    const h = { "Content-Type": "application/json", Accept: "application/json" };
    if (this.apiKey) h["X-API-Key"] = this.apiKey;
    return h;
  }

  async _request(method, path, body) {
    const res = await fetch(this.baseUrl + path, {
      method,
      headers: this._headers(),
      body: body !== undefined ? JSON.stringify(body) : undefined,
    });
    const data = await res.json().catch(() => ({}));
    if (!res.ok) {
      const msg = data?.detail ?? data?.error?.message ?? res.statusText;
      throw new PromptSentinelError(res.status, msg);
    }
    return data;
  }

  // ── public API ───────────────────────────────────────────────────────────

  /**
   * Run a guard scan.
   * @param {string} input           The text to scan (required)
   * @param {string} [output]        LLM output to include in scan
   * @param {string} [context]       Additional context
   * @param {object} [extra]         Extra body fields (e.g. { policy: { block_pii: false } })
   * @returns {Promise<object>}
   */
  async guardScan(input, output, context, extra = {}) {
    const body = { input, ...extra };
    if (output !== undefined) body.output = output;
    if (context !== undefined) body.context = context;
    return this._request("POST", "/guard/scan", body);
  }

  /**
   * Return the current billing-period usage summary.
   * @returns {Promise<object>}
   */
  async getUsageSummary() {
    return this._request("GET", "/usage/summary");
  }

  /**
   * Return the organisation trust score / maturity index.
   * @param {number} [days=30] Look-back window in days
   * @returns {Promise<object>}
   */
  async getTrustScore(days = 30) {
    return this._request("GET", `/analytics/trust-score?days=${days}`);
  }
}
